app/src/main/java/unipd/se18/ocrcamera/SearchResultActivity.java

```diff
@@ -0,0 +1,407 @@
+         * letters
+         */
+        private int lastSize;
+
+        /**
+         * Stores the last dropdown list to save computation time
+         */
+        private List<String> lastList;
+
+        private IngredientsFilter() {
+            super();
+            lastSize = 0;
+            lastList = getFullIngredientsList();
+        }
+
+        @Override
+        protected FilterResults performFiltering(CharSequence constraint) {
+            //this function is called every time the user type anything inside the
+            //autocomplete textview; the constraint is what the user typed so far
+
+            FilterResults results = new FilterResults();
+
+            //without a constraint, show empty suggestions
+            if(constraint == null || constraint.length() == 0) {
+                results.count = 0;
+                return  results;
+            }
+
+            // if the constraint is more restricting than the last, use as full
+            // list the last one and save time
+            List<String> fullList = null;
+            if(constraint.length() - lastSize > 0) {
+                fullList = lastList;
+            } else {
+                // otherwise get a copy of the full ingredients list
+                fullList = getFullIngredientsList();
+            }
+
+            //convert the constraint to upper case as the ingredients from the db use
+            //upper case only
+            final String stringConstraint = constraint.toString().toUpperCase();
+
+            //add to the suggestions any ingredient that contains the constraint
+            ArrayList<String> filteredList = new ArrayList<>();
+            for(String ingredient : fullList) {
+                if(ingredient.contains(stringConstraint)) {
+                    filteredList.add(ingredient);
+                }
+            }
+
+            //sort suggestions (if a suggestions starts with the constraint then show
+            //it before)
+            Collections.sort(filteredList, new Comparator<String> () {
+                @Override
+                public int compare(String a, String b) {
+
+                    //if both strings starts with the constraint, then show the shorter before
+                    if(a.startsWith(stringConstraint) && b.startsWith(stringConstraint))
+                        return a.length() - b.length();
+
+                    //if one of the strings starts with the constraint, then show it before
+                    //using a big value
+                    if(a.startsWith(stringConstraint))
+                        return -100;
+                    if(b.startsWith(stringConstraint))
+                        return +100;
+
+                    String[] subIngredientsA = a.split(" ");
+                    String[] subIngredientsB = b.split(" ");
+
+                    //if any of the substrings starts with the constraint, show the full string
+                    //before
```

Comment:
I would have preferred for-each instead of the traditional for

Reply:
If you inspect the for loop, you can see it starts from i=1 as I don't need to check subIngredients[0]; this is why i used a standard for loop.